### PR TITLE
[SPARK-21150][SQL] Persistent view stored in Hive metastore should be case preserving

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is a regression in Spark 2.2. In Spark 2.2, we introduced a new way to resolve persisted view: https://issues.apache.org/jira/browse/SPARK-18209 , but this makes the persisted view non case-preserving because we store the schema in hive metastore directly. We should follow data source table and store schema in table properties.

## How was this patch tested?

new regression test

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -224,39 +224,36 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
...
+    // specify location for managed table. And in [[CreateDataSourceTableAsSelectCommand]] we have
+    // to create the table directory and write out data before we create this table, to avoid
+    // exposing a partial written table.
+    val needDefaultTableLocation = tableDefinition.tableType == MANAGED &&
+      tableDefinition.storage.locationUri.isEmpty
+
+    val tableLocation = if (needDefaultTableLocation) {
+      Some(CatalogUtils.stringToURI(defaultTablePath(tableDefinition.identifier)))
     } else {
-      // Ideally we should not create a managed table with location, but Hive serde table can
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: this change is actually small, just rearrange the if-else to make view and hive serde table fall into the same branch.

**Comment by cloud-fan** [comment]
cc @gatorsmile @viirya

**Comment by viirya** [comment]
LGTM

**Comment by viirya** [comment]
Oh. I just thought there might be an issue when doing `CREATE OR REPLACE VIEW viewName AS SELECT ...` in `CreateViewCommand`.

The `CatalogTable` used to replace existing view is also coming from `prepareTable`. It doesn't contain the properties. And in `alterTable` we don't explicitly add the properties for `VIEW` too.

So once we replace an existing view, the properties will not be there anymore.

**Comment by cloud-fan** [comment]
@viirya good catch! fixed

**Review by gatorsmile — APPROVED** [review]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.2

**Pull request closed without merging**